src/chain/chain-kernels.cu

```diff
@@ -20,17 +20,9 @@
 #include <cfloat>
 #include "chain/chain-kernels-ansi.h"
 
-
-
 template <typename Real>
 __device__ inline void atomic_add(Real* address, Real value) {
-  Real old = value;
-  Real ret = atomicExch(address, 0.0f);
-  Real new_old = ret + old;
-  while ((old = atomicExch(address, new_old)) != 0.0f) {
-    new_old = atomicExch(address, 0.0f);
-    new_old += old;
-  }
+  atomicAdd(address, value);
```

Comment:
oh OK, fine.

On Tue, Sep 13, 2016 at 10:31 PM, Daniel Galvez <notifications@github.com>
wrote:

> In src/chain/chain-kernels.cu
> <https://github.com/kaldi-asr/kaldi/pull/1025#discussion_r78688088>:
>
> >  template <typename Real>
> >  __device__ inline void atomic_add(Real* address, Real value) {
> > -  Real old = value;
> > -  Real ret = atomicExch(address, 0.0f);
> > -  Real new_old = ret + old;
> > -  while ((old = atomicExch(address, new_old)) != 0.0f) {
> > -    new_old = atomicExch(address, 0.0f);
> > -    new_old += old;
> > -  }
> > +  atomicAdd(address, value);
>
> Yes, it's a little odd. There are subtle details here. My main concern was
> that: since compute capability 6.x (pascal architecture) has an
> implementation of atomicAdd() for double precision floating point, while no
> previous compute capabilities do, changing the name atomic_add to atomicAdd
> for the double instantiation of the template might cause that symbol name
> to conflict with the built-in double precision atomicAdd() library
> (template) function for people using nvcc 8.0 (first nvcc to support
> pascal), causing the build to fail. Note this could be solved by using the
> preprocessor to check the version of nvcc being used to define the custom
> atomicAdd for doubles only if nvcc is less than 8.0. But since both my
> local computer and CLSP have nvcc 7.5 (not 8.0), I couldn't test this
> easily, so I went with the quick odd code change you see now.
> … <#m_-947055387789562811_>
> On Tue, Sep 13, 2016 at 10:19 PM, Daniel Povey ***@***.***> wrote: In
> src/chain/chain-kernels.cu <#1025 (comment)
> <https://github.com/kaldi-asr/kaldi/pull/1025#discussion_r78687401>>: >
> template <typename Real> > __device__ inline void atomic_add(Real* address,
> Real value) { > - Real old = value; > - Real ret = atomicExch(address,
> 0.0f); > - Real new_old = ret + old; > - while ((old = atomicExch(address,
> new_old)) != 0.0f) { > - new_old = atomicExch(address, 0.0f); > - new_old
> += old; > - } > + atomicAdd(address, value); hm. actually... doesn't this
> rather remove the point of having the atomic_add function? You could remove
> the function entirely, and just call atomicAdd(). — You are receiving this
> because you authored the thread. Reply to this email directly, view it on
> GitHub <https://github.com/kaldi-asr/kaldi/pull/1025/files/
> 6f20b39753e858053f73cf724ab09eebb846dae7#r78687401>, or mute the thread <
> https://github.com/notifications/unsubscribe-auth/AEi_UOUiL00RbzMHM_
> kmKhVvbl0i0Vw6ks5qp4PkgaJpZM4J8Zni> .
> -- Daniel Galvez
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1025/files/6f20b39753e858053f73cf724ab09eebb846dae7#r78688088>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVu0rLEaZQRDLLUTHlJ4drJXYhbHfSks5qp4aagaJpZM4J8Zni>
> .
>